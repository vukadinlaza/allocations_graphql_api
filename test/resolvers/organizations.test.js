const { gql } = require("apollo-server");
const { ObjectId } = require("mongodb");
const { getDB } = require("../../src/mongo");
const { FUND_DEAL_ID } = require("../fixtures/deals");
const { ORGANIZATION_ID } = require("../fixtures/organizations");
const { LOGGED_IN_USER_ID, FUND_ADMIN_USER_ID } = require("../fixtures/users");
const { describeWithServer, testError } = require("../setup");

const GET_ORGANIZATION = gql`
  query GetOrganization($slug: String!) {
    organization(slug: $slug) {
      _id
      deals {
        _id
      }
      deal(_id: "${FUND_DEAL_ID}") {
        company_name
      }
      n_deals
      investors {
        _id
      }
      investments {
        _id
      }
      adminInvites {
        to
      }
      approved
    }
  }
`;

const GET_ORGANIZATION_MEMBERS = gql`
  query GetOrganizationMembers($slug: String!) {
    organizationMembers(slug: $slug) {
      _id
    }
  }
`;

const GET_PAGINATED_ORGANIZATIONS = gql`
  query GetPaginatedOrganizations($pagination: PaginationInput!) {
    pagOrganizations(pagination: $pagination) {
      count
      organizations {
        _id
      }
    }
  }
`;

const GET_OVERVIEW_DATA = gql`
  query GetOverviewData($slug: String!) {
    overviewData(slug: $slug)
  }
`;

const CREATE_ORGANIZATION = gql`
  mutation CreateOrganization($org: OrganizationInput!) {
    createOrganization(organization: $org) {
      name
      slug
      approved
      admins {
        _id
      }
    }
  }
`;

const UPDATE_ORGANIZATION = gql`
  mutation UpdateOrganization($org: OrganizationInput!) {
    updateOrganization(organization: $org) {
      name
      slug
      approved
    }
  }
`;

const ADD_ORGANIZATION_MEMBERSHIP = gql`
  mutation AddOrganizationMembership($slug: String!, $user_id: String!) {
    addOrganizationMembership(slug: $slug, user_id: $user_id) {
      _id
    }
  }
`;

const REVOKE_ORGANIZATION_MEMBERSHIP = gql`
  mutation RevokeOrganizationMembership($slug: String!, $user_id: String!) {
    revokeOrganizationMembership(slug: $slug, user_id: $user_id) {
      _id
    }
  }
`;

describeWithServer(
  "Organizations",
  ({
    executeOperationAsAdmin,
    executeOperationAsFundAdmin,
    executeOperationAsInvestor,
    executeOperationAsLoggedIn,
    executeOperation,
  }) => {
    describe("GET_ORGANIZATION", () => {
      test("admin users can get an organization", async () => {
        expect(
          await executeOperationAsAdmin({
            query: GET_ORGANIZATION,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });

      test("fund admin users can get an organization", async () => {
        expect(
          await executeOperationAsFundAdmin({
            query: GET_ORGANIZATION,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });

      testError("non fund admin users cannot get an organization", async () => {
        expect(
          await executeOperationAsInvestor({
            query: GET_ORGANIZATION,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });
    });

    describe("GET_ORGANIZATION_MEMBERS", () => {
      test("admin users can get all orginization members", async () => {
        expect(
          await executeOperationAsAdmin({
            query: GET_ORGANIZATION_MEMBERS,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });

      testError(
        "non admin users cannot get all orginization members",
        async () => {
          expect(
            await executeOperationAsFundAdmin({
              query: GET_ORGANIZATION_MEMBERS,
              variables: { slug: "cool-fund" },
            })
          ).toMatchSnapshot();
        }
      );
    });

    describe("GET_PAGINATED_ORGANIZATIONS", () => {
      test("admin users can get paginated organization results", async () => {
        expect(
          await executeOperationAsAdmin({
            query: GET_PAGINATED_ORGANIZATIONS,
            variables: { pagination: { pagination: 2, currentPage: 0 } },
          })
        ).toMatchSnapshot();
      });

      testError(
        "non-admin users cannot get paginated organization results",
        async () => {
          expect(
            await executeOperationAsFundAdmin({
              query: GET_PAGINATED_ORGANIZATIONS,
              variables: { pagination: { pagination: 2, currentPage: 0 } },
            })
          ).toMatchSnapshot();
        }
      );
    });

    describe("GET_OVERVIEW_DATA", () => {
      test("logged in users can get overview data", async () => {
        expect(
          await executeOperationAsLoggedIn({
            query: GET_OVERVIEW_DATA,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });

      // TODO: skipped because of permissions issues
      test.skip("non logged in users can get overview data", async () => {
        expect(
          await executeOperation({
            query: GET_OVERVIEW_DATA,
            variables: { slug: "cool-fund" },
          })
        ).toMatchSnapshot();
      });
    });

    describe("CREATE_ORGANIZATION", () => {
      test("admin users can create organizations", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsAdmin({
            query: CREATE_ORGANIZATION,
            variables: {
              org: {
                name: "Test Org",
                slug: "test-org",
                approved: true,
              },
            },
          })
        ).toMatchSnapshot();

        expect(
          await db.organizations.findOne({ slug: "test-org" })
        ).toMatchSnapshot({
          _id: expect.any(ObjectId),
          created_at: expect.any(Number),
        });
      });

      testError("non-admin users can create organizations", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsLoggedIn({
            query: CREATE_ORGANIZATION,
            variables: {
              org: {
                name: "Test Org",
                slug: "test-org",
                approved: true,
              },
            },
          })
        ).toMatchSnapshot();

        expect(
          await db.organizations.findOne({ slug: "test-org" })
        ).toMatchSnapshot();
      });
    });

    describe("UPDATE_ORGANIZATION", () => {
      test("fund admin users can update an organization", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsFundAdmin({
            query: UPDATE_ORGANIZATION,
            variables: {
              org: {
                _id: ORGANIZATION_ID,
                slug: "cool-org",
                approved: false,
                name: "OK Org",
              },
            },
          })
        ).toMatchSnapshot();

        expect(
          await db.organizations.findOne({ slug: "cool-org" })
        ).toMatchSnapshot({
          updated_at: expect.any(Number),
        });
      });

      testError(
        "non fund admin users cannot update an organization slug",
        async () => {
          const db = await getDB();

          expect(
            await executeOperationAsInvestor({
              query: UPDATE_ORGANIZATION,
              variables: {
                org: {
                  _id: ORGANIZATION_ID,
                  slug: "my-fund",
                },
              },
            })
          ).toMatchSnapshot();

          expect(
            await db.organizations.findOne({ slug: "cool-fund" })
          ).toMatchSnapshot();
        }
      );
    });

    describe("ADD_ORGANIZATION_MEMBERSHIP", () => {
      test("admin users can add users to an organization", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsAdmin({
            query: ADD_ORGANIZATION_MEMBERSHIP,
            variables: { slug: "cool-fund", user_id: LOGGED_IN_USER_ID },
          })
        ).toMatchSnapshot();

        expect(
          await db.users.findOne({ _id: ObjectId(LOGGED_IN_USER_ID) })
        ).toMatchSnapshot();
      });

      testError(
        "non admin users cannot add users to an organization",
        async () => {
          const db = await getDB();

          expect(
            await executeOperationAsFundAdmin({
              query: ADD_ORGANIZATION_MEMBERSHIP,
              variables: { slug: "cool-fund", user_id: LOGGED_IN_USER_ID },
            })
          ).toMatchSnapshot();

          expect(
            await db.users.findOne({ _id: ObjectId(LOGGED_IN_USER_ID) })
          ).toMatchSnapshot();
        }
      );
    });

    describe("REVOKE_ORGANIZATION_MEMBERSHIP", () => {
      test("admin users can remove users from an organization", async () => {
        const db = await getDB();

        expect(
          await executeOperationAsAdmin({
            query: REVOKE_ORGANIZATION_MEMBERSHIP,
            variables: { slug: "cool-fund", user_id: FUND_ADMIN_USER_ID },
          })
        ).toMatchSnapshot();

        expect(
          await db.users.findOne({ _id: ObjectId(FUND_ADMIN_USER_ID) })
        ).toMatchSnapshot();
      });

      testError(
        "non admin users cannot remove users from an organization",
        async () => {
          const db = await getDB();

          expect(
            await executeOperationAsLoggedIn({
              query: ADD_ORGANIZATION_MEMBERSHIP,
              variables: { slug: "cool-fund", user_id: FUND_ADMIN_USER_ID },
            })
          ).toMatchSnapshot();

          expect(
            await db.users.findOne({ _id: ObjectId(FUND_ADMIN_USER_ID) })
          ).toMatchSnapshot();
        }
      );
    });
  }
);
