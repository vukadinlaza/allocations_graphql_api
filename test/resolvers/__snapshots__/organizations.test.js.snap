// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organizations ADD_ORGANIZATION_MEMBERSHIP admin users can add users to an organization 1`] = `
Object {
  "data": Object {
    "addOrganizationMembership": Object {
      "_id": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations ADD_ORGANIZATION_MEMBERSHIP admin users can add users to an organization 2`] = `
Object {
  "_id": "61142198824e2509714fd54a",
  "email": "altinvestor@allocations.com",
  "first_name": "Luke",
  "last_name": "Skywalker",
  "organizations": Array [],
  "organizations_admin": Array [
    "61119f11beeab561a48939d3",
  ],
}
`;

exports[`Organizations ADD_ORGANIZATION_MEMBERSHIP non admin users cannot add users to an organization 1`] = `
Object {
  "data": Object {
    "addOrganizationMembership": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations ADD_ORGANIZATION_MEMBERSHIP non admin users cannot add users to an organization 2`] = `
Object {
  "_id": "61142198824e2509714fd54a",
  "email": "altinvestor@allocations.com",
  "first_name": "Luke",
  "last_name": "Skywalker",
  "organizations": Array [],
}
`;

exports[`Organizations CREATE_ORGANIZATION admin users can create organizations 1`] = `
Object {
  "data": Object {
    "createOrganization": Object {
      "admins": null,
      "approved": true,
      "name": "Test Org",
      "slug": "test-org",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations CREATE_ORGANIZATION admin users can create organizations 2`] = `
Object {
  "_id": Any<ObjectId>,
  "approved": true,
  "created_at": Any<Number>,
  "name": "Test Org",
  "slug": "test-org",
}
`;

exports[`Organizations CREATE_ORGANIZATION non-admin users can create organizations 1`] = `
Object {
  "data": Object {
    "createOrganization": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations CREATE_ORGANIZATION non-admin users can create organizations 2`] = `null`;

exports[`Organizations GET_ORGANIZATION admin users can get an organization 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "_id": "61119f11beeab561a48939d3",
      "adminInvites": Array [],
      "approved": true,
      "deal": Object {
        "company_name": "test-deal",
      },
      "deals": Array [
        Object {
          "_id": "601c54cbe40e5900230f505d",
        },
        Object {
          "_id": "61143ac37db3b3619db1df42",
        },
      ],
      "investments": Array [
        Object {
          "_id": "6111bb42a4991d5a547352b6",
        },
        Object {
          "_id": "6111bbe6ad565e63b254d0df",
        },
        Object {
          "_id": "61143bc4aa6f42619d924309",
        },
      ],
      "investors": Array [
        Object {
          "_id": "611421d64bf8d10b94e61606",
        },
      ],
      "n_deals": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_ORGANIZATION fund admin users can get an organization 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "_id": "61119f11beeab561a48939d3",
      "adminInvites": Array [],
      "approved": true,
      "deal": Object {
        "company_name": "test-deal",
      },
      "deals": Array [
        Object {
          "_id": "601c54cbe40e5900230f505d",
        },
        Object {
          "_id": "61143ac37db3b3619db1df42",
        },
      ],
      "investments": Array [
        Object {
          "_id": "6111bb42a4991d5a547352b6",
        },
        Object {
          "_id": "6111bbe6ad565e63b254d0df",
        },
        Object {
          "_id": "61143bc4aa6f42619d924309",
        },
      ],
      "investors": Array [
        Object {
          "_id": "611421d64bf8d10b94e61606",
        },
      ],
      "n_deals": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_ORGANIZATION non fund admin users cannot get an organization 1`] = `
Object {
  "data": Object {
    "organization": null,
  },
  "errors": Array [
    [GraphQLError: org query throw],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_ORGANIZATION_MEMBERS admin users can get all orginization members 1`] = `
Object {
  "data": Object {
    "organizationMembers": Array [
      Object {
        "_id": "61142198824e2509714fd54e",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_ORGANIZATION_MEMBERS non admin users cannot get all orginization members 1`] = `
Object {
  "data": Object {
    "organizationMembers": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_OVERVIEW_DATA logged in users can get overview data 1`] = `
Object {
  "data": Object {
    "overviewData": Object {
      "AUM": 30000,
      "SPVs": 1,
      "funds": 1,
      "investors": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_OVERVIEW_DATA non logged in users can get overview data 1`] = `
Object {
  "data": Object {
    "overviewData": Object {
      "AUM": 30000,
      "SPVs": 1,
      "funds": 1,
      "investors": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_PAGINATED_ORGANIZATIONS admin users can get paginated organization results 1`] = `
Object {
  "data": Object {
    "pagOrganizations": Object {
      "count": 1,
      "organizations": Array [
        Object {
          "_id": "61119f11beeab561a48939d3",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations GET_PAGINATED_ORGANIZATIONS non-admin users cannot get paginated organization results 1`] = `
Object {
  "data": Object {
    "pagOrganizations": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations REVOKE_ORGANIZATION_MEMBERSHIP admin users can remove users from an organization 1`] = `
Object {
  "data": Object {
    "revokeOrganizationMembership": Object {
      "_id": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations REVOKE_ORGANIZATION_MEMBERSHIP admin users can remove users from an organization 2`] = `
Object {
  "_id": "61142198824e2509714fd54e",
  "email": "fundadmin@allocations.com",
  "first_name": "Warren",
  "last_name": "Buffett",
  "organizations_admin": Array [],
}
`;

exports[`Organizations REVOKE_ORGANIZATION_MEMBERSHIP non admin users cannot remove users from an organization 1`] = `
Object {
  "data": Object {
    "addOrganizationMembership": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations REVOKE_ORGANIZATION_MEMBERSHIP non admin users cannot remove users from an organization 2`] = `
Object {
  "_id": "61142198824e2509714fd54e",
  "email": "fundadmin@allocations.com",
  "first_name": "Warren",
  "last_name": "Buffett",
  "organizations_admin": Array [
    "61119f11beeab561a48939d3",
  ],
}
`;

exports[`Organizations UPDATE_ORGANIZATION fund admin users can update an organization 1`] = `
Object {
  "data": Object {
    "updateOrganization": Object {
      "approved": false,
      "name": "OK Org",
      "slug": "cool-org",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations UPDATE_ORGANIZATION fund admin users can update an organization 2`] = `
Object {
  "_id": "61119f11beeab561a48939d3",
  "approved": false,
  "name": "OK Org",
  "slug": "cool-org",
  "updated_at": Any<Number>,
}
`;

exports[`Organizations UPDATE_ORGANIZATION non fund admin users cannot update an organization slug 1`] = `
Object {
  "data": Object {
    "updateOrganization": null,
  },
  "errors": Array [
    [GraphQLError: permission denied],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Organizations UPDATE_ORGANIZATION non fund admin users cannot update an organization slug 2`] = `
Object {
  "_id": "61119f11beeab561a48939d3",
  "name": "Cool Fund",
  "slug": "cool-fund",
}
`;
